File: src/main/scala/org/anz/codechallenge/validators/PrimaryKeyValidator.scala
package org.anz.codechallenge.validators

import com.google.gson.Gson
import org.anz.codechallenge.schema.{Schema, Metadata, Tag}
import org.apache.spark.sql.DataFrame
import org.apache.spark.sql.functions.col

import scala.io.Source

/**
  * Primary Key validator
  * @param inputMetadata - Input metadata representing metadata info of input file
  * @param file_schema - File schema representing schema of the input file
  * @param tagFile - Tagfile representing he file name and record count
  * @param dataFrame - Input file read as dataframe which needs validation
  */
class PrimaryKeyValidator (
                            inputMetadata: Metadata,
                            file_schema: Schema,
                            tagFile: Tag,
                            dataFrame: DataFrame
                          ) extends Validator {

  /**
    * Perform primary key validation
    * @return - status of validation
    */
  override def validate(): String = {
    println("Performing primary key validation")

    val colnames = file_schema.primary_keys(0).split(",").map(name => col(name.trim))
    val primaryKeyDf = dataFrame.select(colnames:_*)

    val actualCount = primaryKeyDf.distinct().count()
    val expectedCount = tagFile.record_count

    println("input data records count based on primary key "+actualCount)
    println("expected records count based on primary key"+expectedCount)

    val status = if(actualCount == expectedCount)  "0" else "3"

    println("Primary key validation status is "+status)
    status
  }
}

File: src/main/java/org/anz/codechallenge/validators/PrimaryKeyValidator.java
package org.anz.codechallenge.validators;

import org.anz.codechallenge.filedetails.ContentParams;
import org.anz.codechallenge.filedetails.FileContent;
import org.anz.codechallenge.schema.JSONSchema;
import org.anz.codechallenge.schema.Schema;
import org.anz.codechallenge.tags.DelimitedTag;
import org.apache.spark.sql.Column;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;

public class PrimaryKeyValidator implements Validator {

    private FileContent fileContent;

    public PrimaryKeyValidator(FileContent fileContent) {
        this.fileContent = fileContent;
    }

    @Override
    public String validate() {
        System.out.println("Performing primary key validation");

        Schema fileSchema = fileContent.getFileMetadata().getFileSchema();
        boolean result = false;

        String status = "3";

        if(fileSchema instanceof JSONSchema) {
            JSONSchema jsonSchema = (JSONSchema) fileSchema;

            String[] colArr = jsonSchema.getPrimary_keys().get(0).split(",");

            Column[] colnames = new Column[colArr.length];

            for(int i = 0; i < colArr.length; i++) {
                colnames[i] = new Column(colArr[i]);
            }


            Dataset<Row> primaryKeyDf = fileContent.getDataframe().select(colnames);

            long actualCount = primaryKeyDf.distinct().count();
            int expectedCount = ((DelimitedTag)fileContent.getFileMetadata().getTagFile()).getRecord_count();

            System.out.println("input data records count based on primary key "+actualCount);
            System.out.println("expected records count based on primary key"+expectedCount);

            status = (actualCount == expectedCount) ? "0" : "3";
        }

        System.out.println("Primary key validation status is "+status);
        return status;
    }
}
